File: src/main/java/org/veupathdb/lib/blast/field/Strand.java
package org.veupathdb.lib.blast.field;

import java.util.Locale;
import java.util.Objects;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.databind.JsonNode;
import org.veupathdb.lib.blast.util.DefaultingJSONValue;
import org.veupathdb.lib.blast.util.JSONConstructor;

public enum Strand implements DefaultingJSONValue
{
  Both,
  Minus,
  Plus;

  public static final Strand DefaultStrand = Both;

  public String getValue() {
    return name().toLowerCase(Locale.ROOT);
  }

  @Override
  public boolean isDefault() {
    return this == DefaultStrand;
  }

  @Override
  @JsonValue
  public JsonNode toJSON() {
    return JSONConstructor.newText(getValue());
  }

  public static Strand fromString(String value) {
    Objects.requireNonNull(value);

    return switch (value.toLowerCase(Locale.ROOT)) {
      case "both"  -> Both;
      case "minus" -> Minus;
      case "plus"  -> Plus;

      default -> throw new IllegalArgumentException("Unrecognized strand value \"" + value + "\"");
    };
  }

  @JsonCreator
  public static Strand fromJSON(JsonNode js) {
    return fromString(js.textValue());
  }
}

File: src/main/java/org/veupathdb/lib/blast/field/Strand.kt
package org.veupathdb.lib.blast.field

import org.veupathdb.lib.blast.util.JSONConstructor.newText
import org.veupathdb.lib.blast.util.DefaultingJSONValue
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.annotation.JsonCreator
import java.lang.IllegalArgumentException
import java.util.*

enum class Strand : DefaultingJSONValue {
  Both, Minus, Plus;

  val value: String
    get() = name.lowercase()
  override val isDefault: Boolean
    get() = this == DefaultStrand

  @JsonValue
  override fun toJSON(): JsonNode {
    return newText(value)
  }

  companion object {
    @JvmStatic
    val DefaultStrand = Both

    @JvmStatic
    fun fromString(value: String): Strand {
      Objects.requireNonNull(value)
      return when (value.lowercase()) {
        "both"  -> Both
        "minus" -> Minus
        "plus"  -> Plus
        else    -> throw IllegalArgumentException("Unrecognized strand value \"$value\"")
      }
    }

    @JvmStatic
    @JsonCreator
    fun fromJSON(js: JsonNode): Strand {
      return fromString(js.textValue())
    }
  }
}